File: app/src/main/java/io/jeffchang/buttondemo/ui/common/BaseFragment.java
package io.jeffchang.buttondemo.ui.common;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import dagger.android.support.DaggerFragment;

/**
 * The base which Fragments in this project will extend.
 */

public abstract class BaseFragment extends DaggerFragment {

    private FrameLayout parentView;

    private ProgressBarView progressBarView;

    private ErrorView errorView;

    private View mainView;

    protected abstract int setLayoutResourceID();

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        ViewGroup.LayoutParams params = new ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT);

        Context context = getContext();
        if (context != null) {
            parentView = new FrameLayout(context);
            parentView.setLayoutParams(params);

            progressBarView = new ProgressBarView(context);
            progressBarView.setLayoutParams(params);
            progressBarView.setVisibility(View.GONE);
            parentView.addView(progressBarView);

            errorView = new ErrorView(context);
            errorView.setLayoutParams(params);
            errorView.setVisibility(View.GONE);
            parentView.addView(errorView);

            mainView = inflater.inflate(setLayoutResourceID(), container, false);
            parentView.addView(mainView);
        }

        return parentView;
    }

    public <T extends View> T findParentViewById(int resId) {
        return mainView.findViewById(resId);
    };

    public void showProgressBar(String loadingText) {
        mainView.setVisibility(View.GONE);
        progressBarView.setLoadingText(loadingText);
        progressBarView.setVisibility(View.VISIBLE);
    }

    public void showMainContent() {
        progressBarView.setVisibility(View.GONE);
        mainView.setVisibility(View.VISIBLE);
    }

    public void showErrorView(int error, Runnable runnable) {
        runnable.run();
        progressBarView.setVisibility(View.GONE);
        mainView.setVisibility(View.GONE);
        errorView.setVisibility(View.VISIBLE);
    }
}

File: app/src/main/java/io/jeffchang/buttondemo/ui/common/BaseFragment.kt
package io.jeffchang.buttondemo.ui.common

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout

import dagger.android.support.DaggerFragment

/**
 * The base which Fragments in this project will extend.
 */

abstract class BaseFragment : DaggerFragment() {

    private var parentView: FrameLayout? = null

    private var progressBarView: ProgressBarView? = null

    private var errorView: ErrorView? = null

    private var mainView: View? = null

    abstract var layoutResourceID: Int

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val params = ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT)
        context?.let {
            parentView = FrameLayout(it)
            parentView?.layoutParams = params

            progressBarView = ProgressBarView(it)
            progressBarView?.layoutParams = params
            progressBarView?.visibility = View.GONE
            parentView?.addView(progressBarView)

            errorView = ErrorView(it)
            errorView?.layoutParams = params
            errorView?.visibility = View.GONE
            parentView?.addView(errorView)

            mainView = inflater.inflate(layoutResourceID, container, false)
            parentView?.addView(mainView)
        }
        return parentView
    }

    fun showProgressBar(loadingText: String) {
        mainView?.visibility = View.GONE
        errorView?.visibility = View.GONE
        progressBarView?.loadingText = loadingText
        progressBarView?.visibility = View.VISIBLE
    }

    fun showMainContent() {
        progressBarView?.visibility = View.GONE
        errorView?.visibility = View.GONE
        mainView?.visibility = View.VISIBLE
    }

    fun showErrorView(drawableInt: Int?, error: Int, callback: () -> Unit) {
        progressBarView?.visibility = View.GONE
        mainView?.visibility = View.GONE
        errorView?.tryAgainCallback = callback
        errorView?.titleText = error
        errorView?.iconDrawable = drawableInt
        errorView?.visibility = View.VISIBLE
    }
}
